File: binary_trees/dfs.js
// use 

// runtime disadvatnage 

function helper(node, target) {
  let queue = [node]
  while (queue.length) {
    let node = stack.pop()
    if (node.value == target) {
      return true
    }
    if (node.left) {
      stack.push(node.left)
    }
    if (node.right) {
      stack.push(node.right)
    }
  }
  return false
}


function findDirectParent(tree, a, b) {
  let stack = [tree]
  while (stack.length) {
    let node = stack.pop()
    if (node.val == a) { 
      if((node.left && node.left.val == b) || 
         (node.right && node.right.val == b)) {
        return true
      }
    }
    if (node.left) {
      stack.push(node.left)
    }
    if (node.right) {
      stack.push(node.right)
    }
  }
  return false
}



// TreeNode(1, TreeNode(2))
//      3
//    2   4
//  1 0   5 6
//           7

// 2, 4 -> False
// 2, 0 -> True
// 3, 0 -> True

// 2 5 7

// DFS - xxxxxxxxx <- stack
// BFS - xx <- queue

// Tree equivalent of for loops:

// Recursive DFS
// Iterative DFS
// Iterative BFS


// arr = [1, 2, 3]

*/
/*
     3
   2   4
 1 0   5 6
          7

2, 4 -> False
2, 0 -> True <--
3, 0 -> False

*/
function findDirectParent(tree, a, b) {
  let stack = [tree]
  while (stack.length) {
    let node = stack.pop()
    if (node.val == a) { 
      if((node.left && node.left.val == b) || 
         (node.right && node.right.val == b)) {
        return true
      }
    }
    if (node.left) {
      stack.push(node.left)
    }
    if (node.right) {
      stack.push(node.right)
    }
  }
  return false

File: binary_trees/DFS.py
class TreeNode:
    def __init__(self, value = 0, left = None, right = None):
        self.value = value
        self.left = left
        self.right = right

def find_dfs(root: TreeNode, target: int) -> bool:
    # Write your code here. 
    pass

# Test Cases
tree1 = TreeNode(3, TreeNode(29, TreeNode(2)), TreeNode(4, None, TreeNode(2, TreeNode(9))))
print(find_dfs(None, 1), False)
print(find_dfs(tree1, 9), True)
print(find_dfs(tree1, 1), False)
print(find_dfs(tree1, 2), True)
print(find_dfs(TreeNode(1), 2), False)

# ▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
# Iterative Solution
# ▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔
def find_dfs(node: TreeNode, target: int) -> int:
    if not node:
        return False

    stack = [node]    
    while stack:
        curr = stack.pop()
        if curr.value == target:
            return True
        if curr.left:
            stack.append(curr.left)
        if curr.right:
            stack.append(curr.right)
        
    return False

# ▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
# ⬛️ Recursive Solution
# ▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔
def find_dfs_rec(node, target):
    if not node:
        return False

    if node.value == target:
        return True

    return find_dfs_rec(node.left, target) or find_dfs_rec(node.right, target)
